### PR TITLE
Subscription Manager

### PR DESCRIPTION
**Pull request by kimyen**

### CODE REVIEW

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/subscription/DBOSubscriptionDAOImpl.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	@Override
+	public SubscriptionPagedResults getAll(String subscriberId, Long limit,
+			Long offset, SubscriptionObjectType objectType) {
+		ValidateArgument.required(subscriberId, "subscriberId");
+		ValidateArgument.required(limit, "limit");
+		ValidateArgument.required(offset, "offset");
+		SubscriptionPagedResults results = new SubscriptionPagedResults();
+		long count = getSubscriptionCount(subscriberId, objectType);
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/subscription/DBOSubscriptionDAOImpl.java:107]: SQL_CALC_FOUND_ROWS

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/subscription/DBOSubscriptionDAOImpl.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+			results.setResults(new ArrayList<Subscription>(0));
+		}
+		return results;
+	}
+
+	public static String buildGetQuery(Long limit, Long offset, SubscriptionObjectType objectType) {
+		String query = SQL_GET_ALL;
+		query = addCondition(query, objectType);
+		query += LIMIT + limit + OFFSET + offset;
+		return query;
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/subscription/DBOSubscriptionDAOImpl.java:122]: use bind vars for all sql.

File: lib/lib-upload/src/main/java/org/sagebionetworks/upload/discussion/UploadContentToS3DAOImpl.java [hunk 3]
```diff
@@ -91,7 +91,7 @@ public MessageURL getThreadUrl(String key) {
 		GeneratePresignedUrlRequest request = new GeneratePresignedUrlRequest(
 				bucketName, key).withMethod(HttpMethod.GET).withExpiration(
 				new Date(System.currentTimeMillis()
-				+ PRE_SIGNED_URL_EXPIRATION_MS)).withContentType(TEXT_PLAIN_CHARSET_UTF_8);
+				+ PRE_SIGNED_URL_EXPIRATION_MS));
```
Comment (kimyen) [inline lib/lib-upload/src/main/java/org/sagebionetworks/upload/discussion/UploadContentToS3DAOImpl.java]: revert

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AuthorizationManagerImplUnitTest.java [hunk 4]
```diff
@@ -679,4 +679,18 @@ public void testCanReadBenefactorsTrashNonAdmin(){
...
+		assertEquals(AuthorizationManagerUtil.ACCESS_DENIED,
+				authorizationManager.canSubscribe(userInfo, forumId, SubscriptionObjectType.FORUM));
+	}
+
+	@Test
+	public void testCanSubscribeForumAuthorized() {
+		when(mockEntityPermissionsManager.hasAccess(forumId, ACCESS_TYPE.READ, userInfo)).thenReturn(AuthorizationManagerUtil.AUTHORIZED);
+		assertEquals(AuthorizationManagerUtil.AUTHORIZED,
+				authorizationManager.canSubscribe(userInfo, forumId, SubscriptionObjectType.FORUM));
+	}
```
Comment (kimyen) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/AuthorizationManagerImplUnitTest.java:695]: add test for thread

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/discussion/DiscussionThreadManagerImplAutowiredTest.java [hunk 5]
```diff
@@ -95,7 +95,6 @@ public void test() throws Exception {
 		URL url = new URL(messageUrl.getMessageUrl());
 		HttpURLConnection con = (HttpURLConnection) url.openConnection();
 		con.setRequestMethod("GET");
-		con.setRequestProperty("Content-Type", "text/plain; charset=utf-8");
```
Comment (kimyen) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/discussion/DiscussionThreadManagerImplAutowiredTest.java:98]: revert

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/subscription/SubscriptionManagerImplTest.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+		topic.setObjectType(null);
+		manager.create(userInfo, topic);
+	}
+
+	@Test (expected=UnauthorizedException.class)
+	public void testCreateAnonymous(){
+		when(mockAuthorizationManager
+				.canSubscribe(userInfo, forumId, SubscriptionObjectType.FORUM))
+				.thenReturn(AuthorizationManagerUtil.accessDenied(ANONYMOUS_ACCESS_DENIED_REASON));
+		manager.create(userInfo, topic);
```
Comment (kimyen) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/subscription/SubscriptionManagerImplTest.java:85]: remove

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/subscription/SubscriptionManagerImplTest.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	}
+
+	@Test (expected=UnauthorizedException.class)
+	public void testCreateAccessDenied(){
+		when(mockAuthorizationManager
+				.canSubscribe(userInfo, forumId, SubscriptionObjectType.FORUM))
+				.thenReturn(AuthorizationManagerUtil.ACCESS_DENIED);
+		manager.create(userInfo, topic);
```
Comment (kimyen) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/subscription/SubscriptionManagerImplTest.java:93]: remove

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/subscription/SubscriptionManagerImplTest.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+	@Test
+	public void testCreateAuthorized(){
+		when(mockAuthorizationManager
+				.canSubscribe(userInfo, forumId, SubscriptionObjectType.FORUM))
+				.thenReturn(AuthorizationManagerUtil.AUTHORIZED);
+		when(mockDao
+				.create(userId.toString(), forumId, SubscriptionObjectType.FORUM))
+				.thenReturn(sub);
+		assertEquals(sub, manager.create(userInfo, topic));
```
Comment (kimyen) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/subscription/SubscriptionManagerImplTest.java:104]: verify that authorization manager is being called

File: services/repository/src/main/java/org/sagebionetworks/repo/web/controller/DiscussionController.java [hunk 9]
```diff
@@ -194,10 +194,6 @@ public void markThreadAsDeleted(
 	 * the URL to download the file which contains the thread message.
 	 * <br/>
 	 * Target users: anyone who has READ permission to the project.
-	 * <p>
```
Comment (kimyen) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/controller/DiscussionController.java:197]: revert

**Pull request merged**